proposals/NNNN-custom-isolation-checking-for-serialexecutor.md

```diff
@@ -0,0 +1,190 @@
+# Custom isolation checking for SerialExecutor
+
+* Proposal: [SE-NNNN](NNNN-advanced-custom-isolation-checking-for-serialexecutor.md)
+* Author: [Konrad 'ktoso' Malawski](https://github.com/ktoso)
+* Review Manager: ???
+* Status:  **Work in Progress**
+* Implementation: [PR #71172](https://github.com/apple/swift/pull/71172)
+* Review: ???
+
+## Introduction
+
+Swift introduced custom actor executors in [SE-0392: Custom Actor Executors](https://github.com/apple/swift-evolution/blob/main/proposals/0392-custom-actor-executors.md), and ever since allowed further customization of isolation and execution semantics of actors.
+
+This proposal also introduced a family of assertion and assumption APIs which are able to dynamically check the isolation of a currently executing task. These APIs are:
+
+- Asserting isolation context:
+  - [`Actor/assertIsolated(_:file:line:)`](https://developer.apple.com/documentation/swift/actor/assertisolated(_:file:line:))
+  - [`Actor/preconditionIsolated(_:file:line:)`](https://developer.apple.com/documentation/swift/actor/preconditionisolated(_:file:line:))
+  - [`DistributedActor/assertIsolated(_:file:line:)`](https://developer.apple.com/documentation/distributed/distributedactor/preconditionisolated(_:file:line:))
+  - [`DistributedActor/preconditionIsolated(_:file:line:)`](https://developer.apple.com/documentation/distributed/distributedactor/preconditionisolated(_:file:line:))
+- Assuming isolation context, and obtaining an `isolated actor` reference of the target actor
+  - [`Actor/assumeIsolated(_:file:line:)`](https://developer.apple.com/documentation/swift/actor/assumeisolated(_:file:line:))
+  - [`DistributedActor/assumeIsolated(_:file:line:)`](https://developer.apple.com/documentation/distributed/distributedactor/assumeisolated(_:file:line:))
```

Comment:
I don't think you really need this exhaustive list of APIs; you could just say something like ", such as `assumeIsolated` and `preconditionIsolated`".

The break here between sections is awkward.  This is all really a detailed breakdown of the problem and belongs in Motivation; the Introduction is supposed to be more of an "elevator pitch".  I would suggest something like this:

> SE-0392 added support for custom actor executors, but its support is incomplete.  Safety checks like `assumeIsolated` work correctly when code is running on the actor through a `Task`, but they don't work when code is scheduled to run on the actor's executor through some other mechanism.  For example, if an actor uses a serial `DispatchQueue` as its executor, a function dispatched to the queue with `DispatchQueue.async` cannot use `assumeIsolated` to assert that the actor is currently isolated.  This proposal fixes this by allowing executors to provide custom logic for these safety checks.